### PR TITLE
[Merged by Bors] - feat: `apply _ at _` tactic

### PR DESCRIPTION
**Pull request by adamtopaz**
See relevant zulip discussions [here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/MVars.20from.20implicit.20variables/near/394893597) and [here](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/apply.20at/near/394509077)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by adomani** [comment]
I had already looked at this while it was developing.  Maybe, it makes sense to add the example in the Zulip chat as in a test for the tactic?  Or something that is useful for NNG4?

**Comment by adomani** [comment]
Also, I still wonder whether it would be better to use `apply` rather than building the term "by hand" with `mkAppOptM'`.  I think that `apply` would be more robust and might see through some `DefEq`s that the explicit term would not.  I will not have time to see if I can find an example where one method works and the other does not, but I would probably try with something like `apply f.Injective at (f a = f b)`.

File: Mathlib/Tactic/ApplyAt.lean [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+open Lean Meta Elab Tactic Term
+
+/--
+This function is similar to `forallMetaTelescopeReducingUntilDefEq` except that
+it will construct mvars until it reaches one whose type is defeq to the given
+type `t`. It uses `forallMetaTelescopeReducing`.
+-/
+def Lean.Meta.forallMetaTelescopeReducingUntilDefEq
+    (e t : Expr) (kind : MetavarKind := MetavarKind.natural) :
+    MetaM (Array Expr × Array BinderInfo × Expr) := do
```
Comment (PatrickMassot) [inline Mathlib/Tactic/ApplyAt.lean:23]: This function has nothing to do here, it should be in some utility file, ultimately scheduled for move to sdt4.

File: Mathlib/Tactic/ApplyAt.lean [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      throwError m!"Failed to find {← ppExpr t} as the type of a parameter of {← ppExpr e}."
+    mvs := mvs ++ ms
+    bis := bis ++ bs
+    out := tp
+  return (mvs, bis, out)
+
+namespace Mathlib.Tactic
+
+/--
+`apply t at i` will apply a function `t` at `i`.
```
Comment (PatrickMassot) [inline Mathlib/Tactic/ApplyAt.lean:45]: Did you really meant "function" here?

**Review by PatrickMassot — COMMENTED** [review]
Thanks! However, this definitely requires tests.

File: Mathlib/Tactic/ApplyAt.lean (hunk 1 above)
Comment (adamtopaz) [reply Mathlib/Tactic/ApplyAt.lean:23]: I agree. Do you have a suggestion for where it should be moved? Perhaps in `Mathlib/Lean/Meta/Basic`, given that `forallMetaTelescopeReducing` appears in `Lean/Meta/Basic` in core?

Comment (adamtopaz) [reply Mathlib/Tactic/ApplyAt.lean]: Ok, moved to `Mathlib/Lean/Meta/Basic`.

File: Mathlib/Tactic/ApplyAt.lean (hunk 2 above)
Comment (adamtopaz) [reply Mathlib/Tactic/ApplyAt.lean]: I reworded this a bit.

File: Mathlib/Tactic/ApplyAt.lean [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+Explicitly, if `t : α₁ → ⋯ → αᵢ → ⋯ → αₙ` and `i` has type `αᵢ`, then this tactic will add
+metavariables/goals for any terms of `αⱼ` for `j = 1, …, i-1`,
+then replace the type of `i` with `αᵢ₊₁ → ⋯ → αₙ` by applying those metavariables and the
+original `i`.
+-/
+elab "apply" t:term "at" i:ident : tactic => withMainContext do
+  let f ← Term.elabTerm (← `(@$t)) none
+  let some ldecl := (← getLCtx).findFromUserName? i.getId |
+    throwError m!"Identifier {i} not found"
```
[the remark this answers is not in the document]
Comment (adamtopaz) [reply Mathlib/Tactic/ApplyAt.lean:29]: Thanks! Done.

File: test/ApplyAt.lean [hunk 4]
```diff
@@ -0,0 +1,71 @@
+import Mathlib.Tactic.ApplyAt
+import Mathlib.Algebra.Group.Basic
+import Mathlib.Data.Nat.Basic
+import Std.Tactic.GuardExpr
+
+example {α β : Type*} (f : α → β) (a : α) : β := by
+  apply f at a
+  guard_hyp a :ₛ β
+  exact a
+
```
Comment (joneugster) [inline test/ApplyAt.lean:10]:
```suggestion

/-- `apply at` cannot clear mvarid if still used. -/
example {α : Type} (γ : α → Type) (a : α) (f : α → γ a) : γ a := by
  apply f at a
  rename_i a₂
  guard_hyp a :ₛ γ a₂
  exact a

```

Comment (joneugster) [reply test/ApplyAt.lean:10]: (maybe not useful, but I wrote that trying to understand what `mainGoal.tryClear` does.)

**Review by joneugster — APPROVED** [review]
Looks very clean, and the test cases seem extensive!

(I don't fully understand why `mkAppOptM' f (x |>.map fun e => some e))` does something different than `mkAppM' f x)`, but that's unrelated to this PR.)

File: Mathlib/Lean/Meta/Basic.lean [hunk 5]
```diff
@@ -19,3 +19,31 @@ Restore the metavariable context after execution.
...
+/--
+This function is similar to `forallMetaTelescopeReducing` except that
+it will construct mvars until it reaches one whose type is defeq to the given
+type `t`. It uses `forallMetaTelescopeReducing`.
+-/
+def Lean.Meta.forallMetaTelescopeReducingUntilDefEq
+    (e t : Expr) (kind : MetavarKind := MetavarKind.natural) :
+    MetaM (Array Expr × Array BinderInfo × Expr) := do
+  let mut mvs : Array Expr := #[]
+  let mut bis : Array BinderInfo := #[]
```
[the remark this answers is not in the document]
Comment (adamtopaz) [reply Mathlib/Lean/Meta/Basic.lean:34]: Done

File: Mathlib/Lean/Meta/Basic.lean [hunk 6]
```diff
@@ -19,3 +19,31 @@ Restore the metavariable context after execution.
...
+  let (ms, bs, tp) ← forallMetaTelescopeReducing e (some 1) kind
+  unless ms.size == 1 do
+    if ms.size == 0 then throwError m!"Failed: {← ppExpr e} is not the type of a function."
+    else throwError m!"Failed"
+  mvs := ms
+  bis := bs
```
[the remark this answers is not in the document]
Comment (adamtopaz) [reply Mathlib/Lean/Meta/Basic.lean:40]: Thanks, done.

File: Mathlib/Lean/Meta/Basic.lean [hunk 7]
```diff
@@ -19,3 +19,31 @@ Restore the metavariable context after execution.
...
   let mctx ← getMCtx
   try x finally setMCtx mctx
+
+open Lean Meta
+
+/--
+This function is similar to `forallMetaTelescopeReducing` except that
+it will construct mvars until it reaches one whose type is defeq to the given
+type `t`. It uses `forallMetaTelescopeReducing`.
+-/
```
[the remark this answers is not in the document]
Comment (adamtopaz) [reply Mathlib/Lean/Meta/Basic.lean:29]: Agreed. done.

File: Mathlib/Tactic.lean [hunk 8]
```diff
@@ -1,4 +1,5 @@
 import Mathlib.Tactic.Abel
+import Mathlib.Tactic.ApplyAt
```
[the remark this answers is not in the document]
Comment (adamtopaz) [reply Mathlib/Tactic.lean:2]: Done!

File: Mathlib/Tactic/ApplyAt.lean [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+`apply t at i` will use forward reasoning with `t` at the hypothesis `i`.
+Explicitly, if `t : α₁ → ⋯ → αᵢ → ⋯ → αₙ` and `i` has type `αᵢ`, then this tactic will add
+metavariables/goals for any terms of `αⱼ` for `j = 1, …, i-1`,
+then replace the type of `i` with `αᵢ₊₁ → ⋯ → αₙ` by applying those metavariables and the
+original `i`.
+-/
+elab "apply" t:term "at" i:ident : tactic => withMainContext do
+  let f ← Term.elabTerm (← `(@$t)) none
+  let some ldecl := (← getLCtx).findFromUserName? i.getId |
+    throwErrorAt i m!"Identifier {i} not found"
```
[the remark this answers is not in the document]
Comment (adamtopaz) [reply Mathlib/Tactic/ApplyAt.lean:29]: Done.

File: Mathlib/Tactic/ApplyAt.lean [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  let (mvs, bis, tp) ← forallMetaTelescopeReducingUntilDefEq (← inferType f) ldecl.type
+  let mainGoal ← getMainGoal
+  let mainGoal ← mainGoal.tryClear ldecl.fvarId
+  for (m,b) in mvs.zip bis do
```
[the remark this answers is not in the document]
Comment (adamtopaz) [reply Mathlib/Tactic/ApplyAt.lean:33]: done

File: test/ApplyAt.lean (hunk 4 above)
Comment (adamtopaz) [reply test/ApplyAt.lean:10]: I think it's useful. Adding it now.

File: test/ApplyAt.lean [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+example {α β γ : Type*} (f : α → β) (g : γ) (a : α) : β × γ  := by
+  fail_if_success apply f at g
+  apply f at a
+  guard_hyp a :ₛ β
+  exact (a,g)
+
+example {α β : Type*} (a : α) (b : β) : α × β := by
+  fail_if_success apply a at b
+  exact (a,b)
```
[the remark this answers is not in the document]
Comment (adamtopaz) [reply test/ApplyAt.lean:67]: done.

File: test/ApplyAt.lean [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+example {α β : Type*} (a : α) (b : β) : α × β := by
+  fail_if_success apply a at b
+  exact (a,b)
```
[the remark this answers is not in the document]
Comment (adamtopaz) [reply test/ApplyAt.lean:71]: done

[the remark this answers is not in the document]
Comment (adamtopaz) [reply test/ApplyAt.lean]: Added these tests. Thanks!

**Comment by adamtopaz** [comment]
Thanks for the review!

> Looks very clean, and the test cases seem extensive!
> 
> (I don't fully understand why `mkAppOptM' f (x |>.map fun e => some e))` does something different than `mkAppM' f x)`, but that's unrelated to this PR.)

I tried changing to `mkAppM'` and indeed that makes some of the tests fail. I can't say that I fully understand why, and I don't have much time to investigate right now, unfortunately. So I just left it with `mkAppOptM'` for now.

**Comment by joneugster** [comment]
One last thought I just had, it would be awesome if `apply f at a b ⊢` and `apply f at *` were valid syntax, for consistency with other tactics. But I think that could be a follow-up PR.

**Comment by PatrickMassot** [comment]
bors merge

**Pull request closed without merging**